plugins/todo-backend/src/service/AuthorizedTodoService.ts

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  AuthorizeResult,
+  PermissionAuthorizer,
+} from '@backstage/plugin-permission-common';
+import { stringifyEntityRef } from '@backstage/catalog-model';
+import { NotAllowedError } from '@backstage/errors';
+import { catalogEntityReadPermission } from '@backstage/plugin-catalog-common';
+import { ListTodosRequest, ListTodosResponse, TodoService } from './types';
+
+/** @internal */
+export class AuthorizedTodoService implements TodoService {
+  readonly #delegate: TodoService;
+  readonly #authorizer: PermissionAuthorizer;
+
+  constructor(delegate: TodoService, authorizer: PermissionAuthorizer) {
+    this.#delegate = delegate;
+    this.#authorizer = authorizer;
+  }
+
+  async listTodos(
+    req: ListTodosRequest,
+    options?: { token?: string },
+  ): Promise<ListTodosResponse> {
+    // Permissions are only applied when filtering by entity.
+    // In practice there is no other way to list TODOs, but if one is
+    // added in the future it is important to protect that method too.
+    if (req.entity) {
+      const resourceRef = stringifyEntityRef(req.entity);
+      const [response] = await this.#authorizer.authorize(
+        [{ permission: catalogEntityReadPermission, resourceRef }],
+        { token: options?.token },
+      );
+      if (response.result !== AuthorizeResult.ALLOW) {
+        throw new NotAllowedError();
```

Comment:
This seems to allow consumers of the todo api to differentiate between entities that don't exist and entities that they can't see - should we 404? In fact, isn't there an authorized call to the catalog api somewhere inside `delegate.listTodos` which would itself 404 if the entity didn't exist?

Reply:
Oh yeah totally, this entire implementation as is is actually completely useless :grin: any plugin that look up entities in the catalog don't really need to implement any permissions if all they wanna do is just sync up with the catalog read permission.

Tbh I wanted to continue this by playing around with some custom permissions for the TODO plugin, just to see what that's like and if it be enabled optionally or that kinda thing. Still wanna do that but this puts an entire new spin on things :grin:

Regarding the potential leak it's kinda intentional to bring up the debate x) At the same time as I think it would be nice to not be able to query whether a given entity exists, I also think it may be a doomed endeavor that in the end is not worth the UX tradeoff we're making by having users confused by why they're getting a 404. Whether this leaks information also depends on the permission handler right? If `DENY` is returned also for entities that don't exist, then we'd always be 403'ing here, and there wouldn't be a leak. In the end that might be the way to select the desired behavior? by either choosing to return `ALLOW` or `DENY` for missing resources. It'd result in 403's all over the place rather than 404's, but right now I don't see that that would be a problem.

Reply:
> Whether this leaks information also depends on the permission handler right? If DENY is returned also for entities that don't exist, then we'd always be 403'ing here, and there wouldn't be a leak.

Ah yes, you're quite right. After #8551, the permission handler does indeed return DENY for non-existent entities for exactly this reason. And you're totally right - if we want to avoid leaks (which does seem valuable to me but I agree may be optimistic long term), the important thing is a given endpoint responding consistently for non-existent and forbidden entities - it doesn't matter whether they return 403, 404 or 418 😁.

Reply:
Incidentally, one bit that would be interesting to dig into here is token passing - does the frontend consistently pass a token to the todos-backend, and does that backend consistently include the token when making requests to the catalog? If not, the user will be authorized as if they're anonymous.

Reply:
Long-term we'd be using the `FetchApi` everywhere and have that inject tokens if desired, for now the todo frontend is always grabbing a token from the `IdentityApi` to include, although that may be configured to not return a token, it's up to the integrator